### PR TITLE
Revised object adapter default published endpoints

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR changes the computation of the default published endpoints. It computes them as follows:
- for each endpoint configured on the object adapter, compute the actual port (if needed) and replace the host by _serverName_
- then eliminate duplicates

_serverName_ is a configuration property of the object adapter (_Adapter_.ServerName); its default value is the value of the property Ice.ServerName. And the default value for Ice.ServerName is the current host DNS name.

This update makes specifying `0.0.0.0` or `[::0]` as your object adapter endpoint much more practical - very often the default works, and if not, you can simply set ServerName. Previously, you would get IP addresses or you would need to configure PublishedEndpoints. And with `0.0.0.0:0` or `[::0]:0`, you actually can't configure PublishedEndpoints.

This PR also introduces two new restrictions on object adapter endpoints:
- if an IP endpoint specifies a dynamic port (:0), it must be the only endpoint of the object adapter
- if an IP endpoint specifies a dynamic port (:0), it cannot use a DNS name - its host must be an IP address
These restrictions prevent the creation of object adapters with multiple dynamic IP ports.

Note that as a consequence of this update, the somewhat common object adapter endpoint configuration `tcp -h localhost -p 0` is no longer valid, and should be replaced by `tcp -h 127.0.0.1 -p 0` or `tcp -h "[::1]" -p 0`.

### CODE REVIEW

File: csharp/src/Ice/IPEndpoint.cs [hunk 1]
```diff
@@ -210,7 +215,61 @@ protected internal override void AppendOptions(StringBuilder sb, char optionSepa
...
+                            IPEndpoint expanded = Clone(address.ToString(), Port);
+                            expanded._address = address;
+                            return expanded;
+                        });
+                }
+                catch (Exception ex)
+                {
+                    throw new DNSException(Host, ex);
+                }
+            }
```
Comment (pepone) [inline csharp/src/Ice/IPEndpoint.cs:257]: Can we just use the sync version `Dns.GetHostAddresses`?

```suggestion
            if (Address == IPAddress.None) // DNS name
            {
                try
                {
                    return Dns.GetHostAddresses(Host).Select(
                        address =>
                        {
                            IPEndpoint expanded = Clone(address.ToString(), Port);
                            expanded._address = address;
                            return expanded;
                        });
                }
                catch (Exception ex)
                {
                    throw new DNSException(Host, ex);
                }
            }
            else
            {
                return ImmutableArray.Create(this);
            }
```

File: csharp/src/Ice/ObjectAdapter.cs [hunk 2]
```diff
@@ -798,18 +800,34 @@ public void SetPublishedEndpoints(IEnumerable<Endpoint> newEndpoints, Cancellati
...
+                        {
+                            if (endpoints.Count > 1)
+                            {
+                                throw new InvalidConfigurationException(@$"object adapter `{Name
+                                    }': only one endpoint is allowed when a dynamic IP port (:0) is configured");
+                            }
+                            if (endpoints[0] is IPEndpoint ipEndpoint && ipEndpoint.Address == IPAddress.None)
+                            {
+                                throw new InvalidConfigurationException(@$"object adapter `{Name
+                                    }': use an IP address to configure an endpoint with a dynamic port (:0)");
```
Comment (pepone) [inline csharp/src/Ice/ObjectAdapter.cs:813]: Should we detect this earlier when we parse ObjectAdapter endpoints?

File: csharp/src/Ice/OpaqueEndpoint.cs [hunk 3]
```diff
@@ -111,6 +105,13 @@ protected internal override void AppendOptions(StringBuilder sb, char optionSepa
             }
         }
 
+        protected internal override Endpoint Clone(string host) =>
```
Comment (pepone) [inline csharp/src/Ice/OpaqueEndpoint.cs:108]: Can we just throw `InvalidOperationException`, seems this is only used to compute the OA published endpoints and it cannot contain opaque endpoints.

File: csharp/src/Ice/UniversalEndpoint.cs [hunk 4]
```diff
@@ -87,6 +81,12 @@ protected internal override void AppendOptions(StringBuilder sb, char optionSepa
             }
         }
 
+        protected internal override Endpoint Clone(string host) =>
```
Comment (pepone) [inline csharp/src/Ice/UniversalEndpoint.cs:84]: same here I think this can just throw

**Review by pepone — APPROVED** [review]
Looks good, added a few comments, seems we can remove `Endpoint.ExpandIfWildcard` no longer used

**Review by bentoi — APPROVED** [review]

File: csharp/src/Ice/IPEndpoint.cs (hunk 1 above)
Comment (bernardnormier) [reply csharp/src/Ice/IPEndpoint.cs:257]: Yes, fixed.

File: csharp/src/Ice/ObjectAdapter.cs (hunk 2 above)
Comment (bernardnormier) [reply csharp/src/Ice/ObjectAdapter.cs:813]: I think cleaner to keep the parsing code fairly dumb and do this check right after parsing and not during parsing.

File: csharp/src/Ice/IPEndpoint.cs [hunk 5]
```diff
@@ -210,7 +215,61 @@ protected internal override void AppendOptions(StringBuilder sb, char optionSepa
...
+                {
+                    throw ExceptionUtil.Throw(ex.InnerException!);
+                }
+            }
+            else
+            {
+                return ImmutableArray.Create(this);
+            }
+
+            async Task<IEnumerable<Endpoint>> ExpandDnsHostAsync(CancellationToken cancel = default)
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/IPEndpoint.cs:238]: Fixed.

File: csharp/src/Ice/IPEndpoint.cs [hunk 6]
```diff
@@ -30,12 +31,36 @@ option switch
...
+                    {
+                        _address = IPAddress.Parse(Host);
+                    }
+                    catch (FormatException)
+                    {
+                        // Assume it's a DNS name
+                        _address = IPAddress.None;
+                    }
+                }
+                return _address;
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/IPEndpoint.cs:52]: Fixed.

File: csharp/src/Ice/Endpoint.cs [hunk 7]
```diff
@@ -186,14 +187,17 @@ public override string ToString()
...
+        protected internal abstract Endpoint Clone(string host);
+
+        /// <summary>Expands endpoint into separate endpoints for each IP address returned by the DNS resolver.
         /// </summary>
-        /// <param name="publishedEndpoint">TODO.</param>
-        /// <returns>The collection containing the expanded endpoints.</returns>
-        public abstract IEnumerable<Endpoint> ExpandHost(out Endpoint? publishedEndpoint);
+        /// <returns>The expanded endpoints if Host is a DNS name or IP address and an empty collection if this
+        /// endpoint is not usable as an object adapter endpoint.</returns>
+        // TODO: should this be ExpandHostAsync?
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/Endpoint.cs:199]: Yes, for a follow-up PR.

File: csharp/src/Ice/ColocatedEndpoint.cs [hunk 8]
```diff
@@ -39,15 +39,15 @@ internal class ColocatedEndpoint : Endpoint
...
-            throw new NotSupportedException("colocated endpoint can't be used for object adapter endpoints");
-
         public override IEnumerable<Endpoint> ExpandIfWildcard() => new Endpoint[] { this };
 
         protected internal override void AppendOptions(StringBuilder sb, char optionSeparator)
         {
         }
 
+        protected internal override Endpoint Clone(string host) =>
+            throw new NotImplementedException("cannot clone a colocated endpoint");
```
Comment (externl) [inline csharp/src/Ice/ColocatedEndpoint.cs:49]: Is `NotImplementedException` really the correct type. MS doc says:
*
>You might choose to throw a NotImplementedException exception in properties or methods in your own types when the that member is still in development and will only later be implemented in production code. In other words, a NotImplementedException exception should be synonymous with "still in development."

I think `NotSupportedException` is still the most correct here.

File: csharp/src/Ice/OpaqueEndpoint.cs (hunk 3 above)
Comment (externl) [reply csharp/src/Ice/OpaqueEndpoint.cs:108]: I think we should keep the implementation. In the future we may find another use case.

File: csharp/src/Ice/ConnectionFactory.cs [hunk 9]
```diff
@@ -475,31 +475,26 @@ public void Remove(TKey key, TValue value)
 
     internal abstract class IncomingConnectionFactory : IAsyncDisposable
     {
-        internal Endpoint PublishedEndpoint => _publishedEndpoint ?? Endpoint;
+        internal virtual Endpoint PublishedEndpoint => Endpoint;
 
-        internal abstract void Activate();
-
-        internal abstract void UpdateConnectionObservers();
+        private protected Endpoint Endpoint { get; init; }
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/ConnectionFactory.cs:480]: You're right, it's much cleaner with abstract properties. Fixed.

**Review by externl — APPROVED** [review]
Looks fine. It would be nice if usage with loopback addresses was easier. Now you'll have to set `ServerName` if you want a useable (non-relative) proxy.

What does `Ice.ServerName` and `Adapter.ServerName` mean for non IP/host based transports such as `bt`. Would the user be required to set `Adapter.ServerName` in this case?

I wonder if users will be confused by multiple properties with similar sounding names:
- `Adapter.AdapterID`
- `Adapter.ServerName`

File: csharp/src/Ice/ConnectionFactory.cs [hunk 10]
```diff
@@ -551,14 +546,15 @@ public void Remove(Connection connection)
...
-            : base(endpoint, publish)
+        internal AcceptorIncomingConnectionFactory(ObjectAdapter adapter, Endpoint endpoint, string? serverName)
+            : base(endpoint)
         {
             _communicator = adapter.Communicator;
             _adapter = adapter;
             _acceptor = Endpoint.Acceptor(this, _adapter);
 
             Endpoint = _acceptor.Endpoint;
+            PublishedEndpoint = serverName != null ? Endpoint.Clone(serverName) : Endpoint;
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/ConnectionFactory.cs:557]: Fixed as suggested. 
We should clean this up further in a follow-up PR. It's likely we can eliminate IncomingConnectionFactory.PublishedEndpoints entirely.

File: csharp/src/Ice/OpaqueEndpoint.cs (hunk 3 above)
Comment (bernardnormier) [reply csharp/src/Ice/OpaqueEndpoint.cs:108]: Added a TODO.

File: csharp/src/Ice/UniversalEndpoint.cs (hunk 4 above)
Comment (bernardnormier) [reply csharp/src/Ice/UniversalEndpoint.cs:84]: Added a TODO comment.

File: csharp/src/Ice/ColocatedEndpoint.cs (hunk 8 above)
Comment (bernardnormier) [reply csharp/src/Ice/ColocatedEndpoint.cs:49]: Replaced by `=> this` as suggested by Benoit in an other comment.

**Comment by bernardnormier** [comment]
> Looks fine. It would be nice if usage with loopback addresses was easier. Now you'll have to set `ServerName` if you want a useable (non-relative) proxy.

What do you suggest?

> What does `Ice.ServerName` and `Adapter.ServerName` mean for non IP/host based transports such as `bt`. Would the user be required to set `Adapter.ServerName` in this case?

Maybe the rule should be that _serverName_ only applies to endpoint types with a Host that can be resolved using DNS like IP endpoints?

And for all other endpoints (Bluetooth endpoints, Colocated endpoints, Opaque endpoints, Universal endpoints, ...), the published endpoint is simply the original endpoint.

**Comment by externl** [comment]
> > Looks fine. It would be nice if usage with loopback addresses was easier. Now you'll have to set `ServerName` if you want a useable (non-relative) proxy.
> 
> What do you suggest?

Unfortunately, I don't' have any better ideas. Though maybe it would be nice if you could pass the server name in `CreateObjectAdapter`?
> 
> > What does `Ice.ServerName` and `Adapter.ServerName` mean for non IP/host based transports such as `bt`. Would the user be required to set `Adapter.ServerName` in this case?
> 
> Maybe the rule should be that _serverName_ only applies to endpoint types with a Host that can be resolved using DNS like IP endpoints?
> 
> And for all other endpoints (Bluetooth endpoints, Colocated endpoints, Opaque endpoints, Universal endpoints, ...), the default published endpoint is simply the original endpoint.

Sounds good

File: csharp/src/Ice/OpaqueEndpoint.cs (hunk 3 above)
Comment (bernardnormier) [reply csharp/src/Ice/OpaqueEndpoint.cs:108]: Removed TODO and replaced Clone by GetPublishedEndpoint that throws NotSupportedException as suggested by Jose/Joe.

File: csharp/src/Ice/UniversalEndpoint.cs (hunk 4 above)
Comment (bernardnormier) [reply csharp/src/Ice/UniversalEndpoint.cs:84]: Removed TODO and replaced Clone by GetPublishedEndpoint that throws NotSupportedException as suggested by Jose/Joe.

**Pull request merged**